spec.bs

```diff
@@ -430,17 +439,34 @@ To <dfn>start bidding and scoring</dfn> given an [=auction config=] |auctionConf
           1. [=map/Set=] |perSignalsUrlGenerator|[|joiningOrigin|] to « |ig| ».
         1. Otherwise:
           1. [=list/Append=] |ig| to |perSignalsUrlGenerator|[|joiningOrigin|].
+1. Let |top_score|, and |num_top_bids| be 0.
+1. Let |winning_ad_url| be null.
+1. Let |queue| be the result of [=starting a new parallel queue=].
 1. [=map/For each=] |buyer| → |perBuyerGenerator| of |bidGenerators|:
   1. [=map/For each=] |signalsUrl| -> |perSignalsUrlGenerator| of |perBuyerGenerator|:
-    1. TODO: Fetch |signalsUrl| synchronously.
+    1. TODO: Fetch |signalsUrl| asynchronously.
     1. [=map/For each=] |joiningOrigin| -> |groups| of |perSignalsUrlGenerator|:
-      1. [=list/For each=] |ig| of |groups|: 
-        1. Let |biddingScript| be the result of [=fetch bidding or scoring javascript=] with |ig|'s
-          [=interest group/bidding url=].
-        1. If |biddingScript| is an error, [=iteration/continue=].
-        1. TODO: [=Evaluate script=] with [=evaluate script/script=] set to |biddingScript|,
-          [=evaluate script/functionName=] set to "generateBid", and [=evaluate script/arguments=]
-          set to generateBid's arguemnts.
+      1. [=list/For each=] |ig| of |groups|:
+        1. [=parallel queue/enqueue steps|Enqueue the following steps=] to |queue|:
+          1. TODO: create bidder worklet. If failed, [=iteration/continue=].
+          1. Let |biddingScript| be the result of [=fetch a javascript=] with |ig|'s
+            [=interest group/bidding url=]. If |biddingScript| is an error, [=iteration/continue=].
+          1. Let |generatedBid| be the result of running [=evaluate script=] with
+            [=evaluate script/script=] set to |biddingScript|, [=evaluate script/functionName=] set
+            to "generateBid", and [=evaluate script/arguments=] set to generateBid's arguements.
+          1. If |generatedBid|'s [=generate bid output/bid=] is negative or 0,
+            [=iteration/continue=].
+          1. Let |score| be the result of running [=evaluate script=] with
+            [=evaluate script/script=] set to |decisionLogicScript|,
+            [=evaluate script/functionName=] set to "scoreAd", and [=evaluate script/arguments=] set
+            to scoreAd's arguements.
+          1. If |score| is negative or 0, [=iteration/continue=].
+          1. If |score| is greater than |top_score|, then set |top_score| to |score|, set
+            |winning_ad_url| to |generatedBid|'s [=generate bid output/render=], and add 1 to
+            |num_top_bids|.
+          1. Otherwise, if |score| equals to |top_score|, then set |winning_ad_url| to
+            |generatedBid|'s [=generate bid output/render=] with 1 in |num_top_bids| chance.
+1. Return |winning_ad_url|.

```

Comment:
I don't think this is valid as-is. I think we need to figure out if this algorithm should synchronously wait for `winning_ad_url` to be set/returned from the parallel queue (in which case this algorithm should I guess be called in parallel), or whether it should be called from the main thread. My guess is you want to, in this algorithm, synchronously wait for `winning_ad_url` to be set, but then call this algorithm from in parallel (not currently being done) from `runAdAuction()`.

Reply:
yeah we need to wait for `winning_ad_url` to be set. So some key points of the process I wanted to document was:
1. All bidders start their `generateBid()` in parallel. When a `generateBid()` finishes, a callback is invoked to call `scoreAd()` for that bid. The `scoreAd()` of a bid also runs in parallel with other `generateBid()` and `scoreAd()`.
2. After each `scoreAd()`, it checks whether there are more bids to be generated (there is a counter) and whether there are other `scoreAd()` in progress (has another counter). If both counters are 0, it means generate and score bids are complete, then proceed to the next callback which takes `winning_ad_url` as a parameter.

Reply:
@domfarolino, I see two ways to go here:
1. We write the whole spec to as a very synchronous process, i.e. generate one bid, score that bid, update the winning_bid if it's the highest scoring one, then repeat for each bid.  We could also add a couple sentances saying something like "this bid generation step can be run in parallel with other generate and score steps" and "this scoring step can be run in parallel with other generate and score steps", or
2. try and codify in our spec exactly how to paralellize these steps, i.e. define separate processes/agents, and define task queues, and be very specific about how the queues are maintained and which pieces are parallelized.  My worry here is that we'd be over-specifying implementation details (e.g. our max process limit) into the spec and the spec would be overly complicated and become too hard for API users to understand.

Which approach do you recommend?

Reply:
(my preference is for #1 as I feel it's simpler and easier to understand)

Reply:
Hmm, I think both of those options are on two extreme ends of the spectrum, when neither is strictly necessary. We don't have to write the entire specification as one synchronous process — in fact doing so would be incorrect in seriously observable ways, and defeats the purpose of using the Promise machinery. And I think in order to specify that things go "in parallel" and some things happen asynchronously, we don't have to nail down the exact process allocation policies if it isn't observable. Instead what I think we need is to have `runAdAuction()` call `generate and score bids` _in parallel_ and immediately after return the promise _p_.

Then `generate and score bids` can essentially be written synchronously since nothing on the main thread is blocking it. That is, it can set up some counter and run the various `scoreAd()`s (maybe in parallel), and synchronously wait on their result. Once the URL is computed, then I think you'll want `generate and score bids` to queue a task back to the global that `runAdAuction()` to resolve the promise. At a bare minimum I think that's what is needed.

We set up similar infrastructure in FedCM, maybe looking through their spec would be useful https://fedidcg.github.io/FedCM/#ref-for-create-an-identitycredential

Reply:
@domfarolino Updated, PTAL.
Thanks for the example. Besides that, I also referenced to https://html.spec.whatwg.org/multipage/webappapis.html#event-loop-for-spec-authors, but still not confident that I got it right.

Reply:
I think it's not too bad! I think you want to pass the promise (that `runAdAuction()` creates) _into_ generate and scope bids though, so that it doesn't have to create its own promise.